### PR TITLE
PARQUET-341 improve write performance for wide schema sparse data

### PR DESCRIPTION
**Pull request by tsdeng**
In write path, when there are tons of sparse data, most of time is spent on writing nulls.
Currently writing nulls has the same code path as writing values, which is reclusive traverse all the leaves when a group is null.
Due to the fact that when a group is null all the leaves beneath it should be written with null value with the same repetition level and definition level, we can eliminate the recursion call to get the leaves

This PR caches the leaves for each group node. So when a group node is null, their leaves can be flushed with null values directly.

We tested it with a really wide schema on one of our production data. It improves the performance by ~20%

### CODE REVIEW

**Comment by tsdeng** [comment]
@julienledem @isnotinvain could you take a look at this?

File: parquet-column/src/main/java/org/apache/parquet/io/MessageColumnIO.java [hunk 1]
```diff
@@ -172,16 +175,36 @@ public boolean isWritten(int i) {
...
     private final ColumnWriter[] columnWriter;
+    /** maintain a map of a group and all the leaf nodes underneath it. It's used to optimize writing null for a group node
+     * all the leaves can be called directly without traversing the sub tree of the group node */
+    private Map<GroupColumnIO, List<ColumnWriter>>  groupToLeafWriter = new HashMap<GroupColumnIO, List<ColumnWriter>>();
     private final ColumnWriteStore columns;
     private boolean emptyField = true;
 
+    private void buildGroupToLeafWriterMap(PrimitiveColumnIO primitive, ColumnWriter writer) {
+      GroupColumnIO  parent = primitive.getParent();
+      do {
```
Comment (isnotinvain) [inline parquet-column/src/main/java/org/apache/parquet/io/MessageColumnIO.java]: how about:

```
  do {
    getLeafWriters(parent).add(writer);
    parent = parent.getParent();
  } while (parent != null)
```

to make this a little easier to read?

File: parquet-column/src/main/java/org/apache/parquet/io/MessageColumnIO.java [hunk 2]
```diff
@@ -172,16 +175,36 @@ public boolean isWritten(int i) {
...
+        getLeafWriters(parent).add(writer);
+      } while ((parent = parent.getParent()) != null);
+    }
+
+    private List<ColumnWriter> getLeafWriters(GroupColumnIO group) {
```
Comment (isnotinvain) [inline parquet-column/src/main/java/org/apache/parquet/io/MessageColumnIO.java]: tiny bit more efficient:

```
   private List<ColumnWriter> getLeafWriters(GroupColumnIO group) {
     List<ColumnWriter>() writers = groupToLeafWriter.get(group);
     if (writers == null) {
       writers = new ArrayList<ColumnWriter>();
       groupToLeafWriter.put(group, writers);
     }
     return writers;
   }
```

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java [hunk 3]
```diff
@@ -62,6 +62,7 @@
 
   private MessageType requestedSchema;
   private MessageType fileSchema;
+  private MessageColumnIO columnIO;
```
Comment (isnotinvain) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: why does this need to get held on to? I don't see it being used

**Comment by isnotinvain** [comment]
overall looks good, some minor comments

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java (hunk 3 above)
Comment (tsdeng) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java]: This is actually another problem I found during this PR although not related.
Previouslly, columnIO was not a field, and it gets re-parsed/generated for each row group. This is completely unnecessary.

I am changing it to be generated in the initialize method, and it is reused for each row group.

This change might bring back some performance when the schema is really large.

I can separate this in another PR, what do you think? @isnotinvain

**Comment by tsdeng** [comment]
ping @isnotinvain @julienledem

**Comment by isnotinvain** [comment]
I made two suggestions above, I thought you mentioned that you addressed them but I don't see it in the PR -- do you need to push?

**Comment by isnotinvain** [comment]
Also, now that the columnIO is being cached / re-used, do you know if it is stateful and needs any kind of resetting between row groups? I'm not sure our test coverage for cross-row-group issues is great

**Comment by tsdeng** [comment]
Just pushed, it seems the columnIO is a function of schema that does not contain any state. The RecordConsumer is the one that contains state. I tested it with some real data that spread many row groups and it worked fine...@julienledem could you let me know if I am wrong?

**Comment by isnotinvain** [comment]
+1

**Pull request closed without merging**